Automattic/jetpack
# Dismiss licenses activation notice when license is activated.

#### Changes proposed in this Pull Request:

This PR manually dismisses the licenses activation notice when a Jetpack user(not partner) license key is activated successfully.

Asana task: 1201096622142517-as-1201383441218681

Additional note:
- Sets the notice's "Activate product" link to the correct location at `/{site-admin-page}/admin.php?page=jetpack#/license/activation`
- Also fixes a bug where the license activation notice was showing on Atomic sites.  The notice should only show for self-hosted Jetpack sites (if they have an available unattached license key ready to be activated).

#### Jetpack product discussion



Jetpack Licensed-based Checkout PT: p1HpG7-cSg-p2

#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






**Testing Prerequisites:**
- A WPCOM sandbox account.
- [Redux DevTools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd) extension installed in your browser.

**Testing setup:**
- Spin up a site using **this** Jetpack branch.
- Connect Jetpack to your A8C WordPress.com account.
- If you have tested and dismissed this "License activation notice" in the past, you may want to reset the saved notice dismissal info in the `wp_options` table by clicking the `Reset Options (dev only)` link in the footer of the Admin page (Dashboard).  This will clear out any previous times you may have dismissed the notice.
- Your site should not have any Jetpack plan or product subscriptions. (it should just be "Jetpack Free")

**Get a license key:**
- Sandbox `public-api.wordpress.com` and make sure you have the filter, `add_filter( 'jetpack_license_checkout_m2_enabled', '__return_true' );` in `wp-content/mu-plugins/0-sandbox.php`.  (This filter will make sure a license is created during purchase.)
- Go to https://cloud.jetpack.com/pricing
- Select a paid Jetpack product
- Complete checkout.
- On the "Thank you for purchasing [product-name]!" page, just stop there..  (Don't enter a site URL in the input box).
- You can go to Store Admin to view the subscription you just purchased.  You should see your license key in the subscription info. (You'll need it later)
- Now un-sandbox `public-api.wordpress.com` (comment it out in your hosts file). We don't want/need to be sandboxed when testing the Jetpack plugin in the next steps)

**Test the license activation notice and the dismissal of the notice:**
- Back in your Jetpack site, go to the Jetpack Dashboard. ( `/wp-admin/admin.php?page=jetpack#/dashboard` ).
- Verify that you can see the "**You have an inactive product. Activate now**" notice at the top of the page.
- Using Redux DevTools inspect the `state.jetpack.licensing.activationNoticeDismiss` objects.  Make a mental note of the `activationNoticeDismiss.lastDetachedCount` and `activationNoticeDismiss.lastDetachedTime` values. (We will dismiss the notice and observe that these values get updated).
- Dismiss (click the X) the "**You have an inactive product. Activate now**" notice. 
- Verify that the notice disappears and that the `activationNoticeDismiss.lastDetachedCount` value is updated,  containing the current value of `state.jetpack.licensing.userCounts.detached` and that `activationNoticeDismiss.lastDetachedTime` is updated with the current time(UTC).
- Reload the page and verify the activation notice stays dismissed.

**Test that the license activation notice dismisses itself when a license key is successfully activated:**
- First we need to reset the saved notice dismissal info in the `wp_options` table by clicking the `Reset Options (dev only)` link in the footer of the Admin page (Dashboard).  This will clear out the previous time you dismissed the notice, so the notice will show again.
- Reload the dashboard page, the "**You have an inactive product. Activate now**" notice should be showing.
- Click the "Activate now" link inside the notice.
- Verify you are taken to the License Activation page: (`/wp-admin/admin.php?page=jetpack#/license/activation`)
- Back In the Store Admin, copy the license key from the subscription you purchased earlier.
- Using Redux DevTools again, inspect the `state.jetpack.licensing.userCounts` and `state.jetpack.licensing.activationNoticeDismiss` objects.
- In the plugin, on the License Activation page, paste the license key into the input box, and submit.
- Upon successful product license key activation, verify the values of `state.jetpack.licensing.userCounts.detached`, `activationNoticeDismiss.lastDetachedCount` and `activationNoticeDismiss.lastDetachedTime` have been updated.
- Go to the Jetpack Dashboard page ( `/wp-admin/admin.php?page=jetpack#/dashboard` ) and verify the activation notice is no longer showing.
- Click on "My Plan", and verify you can see the new plan(or product) you just activated.

**Test that the notice re-appears again after obtaining a new un-attached license key:**
- Get another license key (Complete the steps under the "**Get a license key**" section described above).
- Go back to your Jetpack site and load the Jetpack Dashboard page.
- Verify you see the "**You have an inactive product. Activate now**" notice again.
